src/modules/zfs/ZfsJob.cpp

```diff
@@ -0,0 +1,326 @@
+                m_mountpoints.append( mountpoint );
+            }
+        }
+    }
+}
+
+bool
+ZfsJob::IsMountpointOverlapping( const QString& targetMountpoint ) const
+{
+    for ( const QString& mountpoint : m_mountpoints )
+    {
+        if ( mountpoint != '/' && targetMountpoint.startsWith( mountpoint ) )
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+
+ZfsResult
+ZfsJob::CreateZpool( QString deviceName, QString poolName, QString poolOptions, bool encrypt, QString passphrase ) const
+{
+    // zfs doesn't wait for the devices so pause for 2 seconds to ensure we give time for the device files to be created
+    QString command;
+    if ( encrypt )
+    {
+        command = "sleep 2 ; echo \"" + passphrase + "\" | zpool create " + poolOptions
+            + " -O encryption=aes-256-gcm -O keyformat=passphrase " + poolName + " " + deviceName;
+    }
+    else
+    {
+        command = "sleep 2 ; zpool create " + poolOptions + " " + poolName + " " + deviceName;
+    }
+
+    // We use a qProcess instead of runCommand so the password will not end up in the logs
+    QProcess process;
+
+    process.setProcessChannelMode( QProcess::MergedChannels );
+    cDebug() << Logger::SubEntry << "Running zpool create";
+
+    process.start( "sh", QStringList() << "-c" << command );
+
+    if ( !process.waitForStarted() )
+    {
+        return { false, tr( "zpool create process failed to start" ) };
+    }
+
+    if ( !process.waitForFinished( 5000 ) )
+    {
+        return { false, tr( "Process for zpool create timed out" ) };
+    }
+
+    QString output = QString::fromLocal8Bit( process.readAllStandardOutput() ).trimmed();
+
+    if ( process.exitStatus() == QProcess::CrashExit )
+    {
+        return { false, tr( "The output from the crash was: " ) + output };
+    }
+
+    auto exitcode = process.exitCode();
+    if ( exitcode != 0 )
+    {
+        cWarning() << "Failed to run zpool create.  The output was: " + output;
+        return { false, tr( "Failed to create zpool on " ) + deviceName };
+    }
+
+    return { true, QString() };
+}
+
+Calamares::JobResult
+ZfsJob::exec()
+{
+    QVariantList partitions;
+    Calamares::GlobalStorage* gs = Calamares::JobQueue::instance()->globalStorage();
+    if ( gs && gs->contains( "partitions" ) && gs->value( "partitions" ).canConvert( QVariant::List ) )
+    {
+        partitions = gs->value( "partitions" ).toList();
+    }
+    else
+    {
+        cWarning() << "No *partitions* defined.";
+        return Calamares::JobResult::internalError( tr( "Configuration Error" ),
+                                                    tr( "No partitions are available for Zfs." ),
+                                                    Calamares::JobResult::InvalidConfiguration );
+    }
+
+    const CalamaresUtils::System* system = CalamaresUtils::System::instance();
+
+    QVariantList poolNames;
+
+    for ( auto& partition : qAsConst( partitions ) )
+    {
+        QVariantMap pMap;
```

Comment:
From here to L176 looks like a function -- ok, I like short functions that tell a story -- that returns `QPair<QString, QString>` of device-name and mount-point. Maybe some more data-structure wrangling is needed there.